pkg/migrate/marketplacemigration.go

```diff
@@ -0,0 +1,367 @@
+func NewMarketplaceMigration(client client, cfg config.CentralConfig, cache cache.Manager) *MarketplaceMigration {
+	return &MarketplaceMigration{
+		client: client,
+		cfg:    cfg,
+		cache:  cache,
+	}
+}
+
+// MarketplaceMigration - used for migrating attributes to subresource
+type MarketplaceMigration struct {
+	client                  client
+	cfg                     config.CentralConfig
+	cache                   agentcache.Manager
+	accessRequestDefinition *mv1a.AccessRequestDefinition
+}
+
+// Migrate -
+func (m *MarketplaceMigration) Migrate(ri *v1.ResourceInstance) (*v1.ResourceInstance, error) {
+	if ri.Kind != mv1a.APIServiceGVK().Kind {
+		return ri, fmt.Errorf("expected resource instance kind to be api service")
+	}
+
+	log.Debugf("migrating marketplace provisioning for service: %s", ri.Name)
+
+	funcs := []migrateFunc{
+		m.updateInst,
+	}
+
+	errCh := make(chan error, len(funcs))
+	wg := &sync.WaitGroup{}
+
+	for _, f := range funcs {
+		wg.Add(1)
+
+		go func(fun migrateFunc) {
+			defer wg.Done()
+
+			err := fun(ri)
+			errCh <- err
+		}(f)
+	}
+
+	wg.Wait()
+	close(errCh)
+
+	for e := range errCh {
+		if e != nil {
+			return ri, e
+		}
+	}
+
+	log.Debugf("finished migrating marketplace provisioning for service: %s", ri.Name)
+
+	return ri, nil
+}
+
+// updateInst gets a list of instances for the service and updates their request definitions.
+func (m *MarketplaceMigration) updateInst(ri *v1.ResourceInstance) error {
+	revURL := m.cfg.GetRevisionsURL()
+
+	q := map[string]string{}
+
+	revs, err := m.client.GetAPIV1ResourceInstancesWithPageSize(q, revURL, 100)
+	if err != nil {
+		return err
+	}
+
+	errCh := make(chan error, len(revs))
+	wg := &sync.WaitGroup{}
+
+	// query for api service instances by reference to a revision name
+	for _, rev := range revs {
+		wg.Add(1)
+
+		go func(r *v1.ResourceInstance) {
+			defer wg.Done()
+
+			q := map[string]string{
+				"query": queryFunc(r.Name),
+			}
+			url := m.cfg.GetInstancesURL()
+			err := m.updateInstResources(url, q, r)
+			errCh <- err
+		}(rev)
+	}
+
+	wg.Wait()
+	close(errCh)
+
+	for e := range errCh {
+		if e != nil {
+			return e
+		}
+	}
+
+	return nil
+}
+
+func (m *MarketplaceMigration) updateInstResources(resourceURL string, query map[string]string, resourceInstance *v1.ResourceInstance) error {
+	resources, err := m.client.GetAPIV1ResourceInstancesWithPageSize(query, resourceURL, 100)
+	if err != nil {
+		return err
+	}
+
+	wg := &sync.WaitGroup{}
+	errCh := make(chan error, len(resources))
+
+	for _, resource := range resources {
+		wg.Add(1)
+
+		go func(ri *v1.ResourceInstance) {
+			defer wg.Done()
+
+			apiSvcInst := mv1a.NewAPIServiceInstance(ri.Name, ri.Metadata.Scope.Name)
+			apiSvcInst.FromInstance(ri)
+
+			specDefintionType := resourceInstance.Spec["definition"].(map[string]interface{})["type"].(string)
+			specDefinitionValue := resourceInstance.Spec["definition"].(map[string]interface{})["value"].(string)
```

Comment:
You might want to check if the type conversion was ok for both of these

Reply:
Done...